src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -878,7 +865,10 @@ DagState uniquifyState(ImmutableArray<StateForCase> cases, ImmutableDictionary<B
                     {
                         // There is a when clause to evaluate.
                         // In case the when clause fails, we prepare for the remaining cases.
-                        var stateWhenFails = state.Cases.RemoveAt(0);
+                        var stateWhenFails = ArrayBuilder<StateForCase>.GetInstance(state.Cases.Count - 1);
+                        for (int i = 1, n = state.Cases.Count; i < n; i++)
+                            stateWhenFails.Add(state.Cases[i]);
```

Comment:
avoided N^2 allocation of arrays.  With a large array, this would then create each array one smaller.  Now this is just an array builder, which can go back to the pool when done.